src/tests/Microsoft.Diagnostics.Monitoring/EndpointInfoSourceTests.cs

```diff
@@ -137,6 +137,76 @@ await using (var execution1 = StartTraceeProcess("LoggerRemoteTest", transportNa
             Assert.Empty(endpointInfos);
         }
 
+        /// <summary>
+        /// Tests that the server endpoint info source can properly enumerate endpoint infos when multiple
+        /// targets connect to it and "disconnect" from it.
+        /// </summary>
+        [Fact]
+        public async Task ServerSourceAddRemoveMultipleConnectionTest()
+        {
+            await using var source = CreateServerSource(out string transportName);
+            source.Start();
+
+            var endpointInfos = await GetEndpointInfoAsync(source);
+            Assert.Empty(endpointInfos);
+
+            const int appCount = 5;
+            RemoteTestExecution[] executions = new RemoteTestExecution[appCount];
+
+            try
+            {
+                // Start all app instances
+                for (int i = 0; i < appCount; i++)
+                {
+                    Task newEndpointInfoTask = source.WaitForNewEndpointInfoAsync(DefaultPositiveVerificationTimeout);
+
+                    executions[i] = StartTraceeProcess("LoggerRemoteTest", transportName);
+
+                    await newEndpointInfoTask;
+
+                    executions[i].SendSignal();
+                }
+
+                await Task.Delay(TimeSpan.FromSeconds(1));
+
+                endpointInfos = await GetEndpointInfoAsync(source);
+
+                Assert.Equal(appCount, endpointInfos.Count());
+
+                for (int i = 0; i < appCount; i++)
+                {
+                    IEndpointInfo endpointInfo = endpointInfos.FirstOrDefault(info => info.ProcessId == executions[i].TestRunner.Pid);
+                    Assert.NotNull(endpointInfo);
+                    Assert.NotNull(endpointInfo.CommandLine);
+                    Assert.NotNull(endpointInfo.OperatingSystem);
+                    Assert.NotNull(endpointInfo.ProcessArchitecture);
+
+                    VerifyConnection(executions[i].TestRunner, endpointInfo);
+                }
+            }
+            finally
+            {
+                _outputHelper.WriteLine("Stopping tracees.");
+
+                int executionCount = 0;
+                for (int i = 0; i < appCount; i++)
+                {
+                    if (null != executions[i])
+                    {
+                        executionCount++;
+                        await executions[i].DisposeAsync();
+                    }
+                }
+                Assert.Equal(appCount, executionCount);
+            }
+
+            await Task.Delay(TimeSpan.FromSeconds(1));
```

Comment:
Same as above

Reply:
This one might be needed, as the other tests are doing this after stopping the test application. What I've observed is if we query for the processes immediately after shutting them down, their pipes are still viable for a short period. So the server will show that they exist even though they really don't. If we wait a bit, then the server has enough time reflect the correct state.

Reply:
I'll keep this Task.Delay as-is and look for a way to determine when the pipes for the processes are finally gone. Maybe I can do an async tight loop to check when they are gone (with a reasonable timeout) and then continue the test so that way we can eliminate the delays.